Catalogue of Things:Capstone Project

RileyManda opened the pull request:
We as a team @rabayarabu | @DianaBeki | @RileyManda have completed the task of building the catalogue of things application according to the [project-requirements](https://github.com/microverseinc/curriculum-professional-skills/blob/main/interview-prep/share-behavioral-questions-videos-with-your-partner.md) 🌱 

### Implementations:

**Group Tasks:**
- Create Item class in a separate .rb file.
- All Item class properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
- Add all methods visible in the diagram.
- Implement methods:
- can_be_archived?() in the Item class
- should return true if published_date is older than 10 years
- otherwise, it should return false
- move_to_archive() in the Item class
- should reuse can_be_archived?() method
- should change the archived property to true if the result of the can_be_archived?() method is true
- should do nothing if the result of the can_be_archived?() method is false
- Create a main.rb file that will serve as your console app entry-point.
- Implement startup actions:
- Present the user with a list of options to perform.
- Let users choose an option.
- If needed, ask for parameters for the option.
- Have a way to quit the app.

**Team member #1**
- Create a Book class in a separate .rb file.
- Create a Label class with an association to the Item class (in a separate .rb file).
- All Book class properties visible in the diagram should be defined and set up in the constructor method.
- All Label class properties visible in the diagram should be defined and set up in the constructor method.
- Implement methods:
- add_item method in the Label class
- should take an instance of the Item class as an input
- should add the input item to the collection of items
- should add self as a property of the item object (by using the correct setter from the item object)
- can_be_archived?() in the Book class
- should override the method from the parent class
- should return true if parent's method returns true OR if cover_state equals to "bad"
- otherwise, it should return false
- Add unit tests for all implemented methods.
- The following options should be available:
- List all books
- List all labels (e.g. 'Gift', 'New')
- Add a book
- All data should be preserved by saving collections in .json files.
- Create a schema.sql file with tables that will be analogical to the structure of the classes that you created:
- books table (add all properties and associations from the parent Item class as table columns)
- labels table

**Team member #2**
- Create MusicAlbum class in a separate .rb file.
- Create Genre class with an association to the Item class (in a separate .rb file).
- All MusicAlbum class properties visible in the diagram should be defined and set up in the constructor method.
- All Genre class properties visible in the diagram should be defined and set up in the constructor method.
- Implement methods:
- add_item method in the Genre class
- should take an instance of the Item class as an input
- should add the input item to the collection of items
- should add self as a property of the item object (by using the correct setter from the item object)
- can_be_archived?() in the MusicAlbum class
- should override the method from the parent class
- should return true if parent's method returns true AND if on_spotify equals true
- otherwise, it should return false
- Add unit tests for all implemented methods.
- The following options should be available:
- List all music albums
- List all genres (e.g 'Comedy', 'Thriller')
- Add a music album
- All data should be preserved by saving collections in .json files.
- Create a schema.sql file with tables that will be analogical to the structure of the classes that you created:
- music_albums table (add all properties and associations from the parent Item class as table columns)
- genres table

**Team member #3** 

- Create a Game class in a separate .rb file.
- Create an Author class with an association to the Item class (in a separate .rb file).
- All Game class properties visible in the diagram should be defined and set up in the constructor method.
- All Author class properties visible in the diagram should be defined and set up in the constructor method.
- Implement methods:
- add_item method in the Author class
- should take an instance of the Item class as an input
- should add the input item to the collection of items
- should add self as a property of the item object (by using the correct setter from the item object)
- can_be_archived?() in the Game class
- should override the method from the parent class
- should return true if parent's method returns true AND if last_played_at is older than 2 years
- otherwise, it should return false
- Add unit tests for all implemented methods.
- The following options should be available:
- List of games
- List all authors (e.g. 'Stephen King')
- Add a game
- All data should be preserved by saving collections in .json files.
- Create a schema.sql file with tables that will be analogical to the structure of the classes that you created:
- games table (add all properties and associations from the parent Item class as table columns)
- authors table 

Please kindly review our implementations and suggest any improvements
Thank you 🙏🏻

RileyManda committed:
Added readme template

RileyManda committed:
Linters|rubocop|dependencies gems setup

RileyManda committed:
Merge pull request #5 from RileyManda/project-setup

Project setup

RileyManda committed:
Created file schema

RileyManda committed:
Created db catalog

RileyManda committed:
Created music_albums schema

RileyManda committed:
Created table genres schema

RileyManda committed:
Items table schema and serialized items table

RileyManda committed:
Updated class attributes

RileyManda committed:
Updated class album,item attrubutes

RileyManda committed:
Created things db

RileyManda committed:
Removed redundant table

RileyManda committed:
Fixed typo in constraint

RileyManda committed:
Merge pull request #26 from RileyManda/feature/schema

Feature/schema

rabayarabu committed:
Create table books

rabayarabu committed:
Create table games

rabayarabu committed:
Create table authors

RileyManda committed:
Merge pull request #33 from RileyManda/feature/schema

Add schema for tables: books, labels, games and authors

RileyManda committed:
Updated table names to match uml diagram naming convention:Updated db name to catalog

rabayarabu committed:
Merge pull request #36 from RileyManda/feature/schema-iterations

Updated table names to match uml diagram naming convention:Updated db…

RileyManda committed:
Created music album class with constructor and its properties

rabayarabu committed:
Merge pull request #37 from RileyManda/feature/music_album

Created music album class

rabayarabu committed:
Add book and label class

DianaBeki committed:
Create an empty game class

RileyManda committed:
Merge pull request #38 from RileyManda/feature/book

Add book and label class

RileyManda committed:
Created genre class with all its properties

RileyManda committed:
Merge pull request #40 from RileyManda/feature/genre

Created genre class with all its properties

RileyManda committed:
Item class params

DianaBeki committed:
Fix linters and add constructor method to game class

RileyManda committed:
Implemented options hash to reduce long lines

RileyManda committed:
Implemented archive checks/condition

DianaBeki committed:
Fix Linters

DianaBeki committed:
Merge pull request #39 from RileyManda/feature/game-class

Create an empty game class

RileyManda committed:
Merge branch 'dev' into feature/item-class

RileyManda committed:
Linter error fixes

RileyManda committed:
Merge pull request #41 from RileyManda/feature/item-class

Feature/item class

DianaBeki committed:
Update documentation

DianaBeki committed:
Create author class and set up constructor method

RileyManda committed:
Implemented self and array method to genre

RileyManda committed:
Updated music album method to override

DianaBeki committed:
Merge pull request #43 from RileyManda/feature/author-class

Create an Author class

rabayarabu committed:
Merge pull request #42 from RileyManda/feature/album-methods

Feature/album methods

RileyManda committed:
Bug fixes|Implemented missing methods for list items

RileyManda committed:
Linter error fix

DianaBeki committed:
Update documentation

RileyManda committed:
Fixed long line and seperated methods into seperate blocks

RileyManda committed:
Merge readme

RileyManda committed:
bug fixes

RileyManda committed:
Updated app.rb

RileyManda committed:
List all genres

RileyManda committed:
List authors

RileyManda committed:
Fixed cyclomatic linter issue be restructuring display item

RileyManda committed:
Updated list of options to match requirements

RileyManda committed:
Merge pull request #45 from RileyManda/feature/main-class-stable

main class-iterations|bug fixes.

rabayarabu committed:
Add label and show list

rabayarabu committed:
Add label when add book

RileyManda committed:
Linter fixes:implemented super and reduced code lines

RileyManda committed:
List all books and labels bug fix

RileyManda committed:
Updated methods for labels and books in main.rb|broke methods into smaller chunks to fix long line linter issue

RileyManda committed:
Merge pull request #46 from RileyManda/feature/main-class-stable

Add label in the book

RileyManda committed:
Display music albums bug fix:missing method

RileyManda committed:
Bug fixes:display game|musicalbum

rabayarabu committed:
Merge pull request #47 from RileyManda/feature/main-class-stable

Feature/main class stable

RileyManda committed:
Created json files genre music_album

RileyManda committed:
Implemented save music_album to json

RileyManda committed:
Save genre to genre.json

RileyManda committed:
Bug fix:save music_album with id

RileyManda committed:
Bug fix:save genre to json file

RileyManda committed:
Updated json structure:album.json

RileyManda committed:
List genres

RileyManda committed:
List all music albums-Bug fix

DianaBeki committed:
Preserve data

RileyManda committed:
Update genre json file

RileyManda committed:
Updated add genre:increment array

RileyManda committed:
save genres and albums bug fix

RileyManda committed:
Updated add music album

RileyManda committed:
Bug trace...wip

RileyManda committed:
Implemented add music and genre incrementally

RileyManda committed:
Updated add music-included genre id to array

RileyManda committed:
Code cleanup

RileyManda committed:
Bug fix:Fixed wrong no. of arguments in method

RileyManda committed:
Updated music data

RileyManda committed:
Music album methods-restructure

RileyManda committed:
Cyclomatic dep fix

RileyManda committed:
Updated method name with no prefix naming convention

RileyManda committed:
Lin lenght|ignore

RileyManda committed:
Test data:

RileyManda committed:
Linter fixes

RileyManda committed:
music-data

RileyManda committed:
Testing terminal application data gen

RileyManda committed:
Long line fix

RileyManda committed:
Fixed long line issue with method restructuring

RileyManda committed:
Merge pull request #48 from RileyManda/feature/preserve-data-music

Feature/preserve data music

RileyManda committed:
Implemented add music album to items collection

RileyManda committed:
Bug fix load items

RileyManda committed:
Fixd issue with incrementing music albums data to items.json

RileyManda committed:
Bug fix: show genre for each music album

RileyManda committed:
Renamed overused variable

RileyManda committed:
Genre arguments number fix

RileyManda committed:
Genre bug fix

RileyManda committed:
Linter fixes

RileyManda committed:
Removed debug code

rabayarabu committed:
Store books

RileyManda committed:
Merge pull request #49 from RileyManda/feature/preserve-data-music

Preserve data-music

rabayarabu committed:
Fix spelling mistake

rabayarabu committed:
Fixed arguments error

rabayarabu committed:
Load books

RileyManda committed:
Fixed linter errors

RileyManda committed:
Merge branch 'dev' into store-book

RileyManda committed:
Fixed linter issue

DianaBeki committed:
Preserve data

DianaBeki committed:
Fix linters

RileyManda committed:
Fixed repeating title prompt and fixed label not visible in list books

RileyManda committed:
Linter fixes

rabayarabu committed:
Merge pull request #50 from RileyManda/store-book

Store book

rabayarabu committed:
Test book & label logic

RileyManda committed:
Expect album entries to have unieuw id:test

RileyManda committed:
Expect album name:test

RileyManda committed:
Expect:can be archived:test

RileyManda committed:
Expects the genre attribute to be set correctly

RileyManda committed:
Expects can_be_archived? to return the correct value

rabayarabu committed:
Merge pull request #52 from RileyManda/rspec/book

Test book & label logic

RileyManda committed:
Moved test file to spec directory

RileyManda committed:
Write music-album to json tests

RileyManda committed:
Linter error fixes

RileyManda committed:
Should save genres to the JSON:test

DianaBeki committed:
Fix linters

DianaBeki committed:
Add game

RileyManda committed:
Should load genres from existing files:genre.json

RileyManda committed:
Should load music albums correctly from an existing:music.json-test

RileyManda committed:
Should return empty array if file does not exist

rabayarabu committed:
Merge branch 'dev' into feature/preserve-data-games

RileyManda committed:
Bug fix-listing genres

RileyManda committed:
Merge pull request #53 from RileyManda/feature/music-album-tests

Feature/music album tests

DianaBeki committed:
Preseve data

DianaBeki committed:
Preserve data

rabayarabu committed:
Merge pull request #54 from RileyManda/feature/preserve-data-games

Preserving Data

rabayarabu committed:
Fixed label bug

DianaBeki committed:
Unit test

DianaBeki committed:
Run test for my game class

rabayarabu committed:
Merge pull request #55 from RileyManda/fix/book-label

Fixed label bug

DianaBeki committed:
Fix linters

DianaBeki committed:
Fix linters

RileyManda committed:
Removed list all items option and updated options

rabayarabu committed:
Merge pull request #56 from RileyManda/rspec/games

Unit test for Game class

RileyManda committed:
Bug fix:music archive and spotify boolean inputs

RileyManda committed:
Linter fixes

rabayarabu committed:
Merge pull request #57 from RileyManda/feature/options

Feature/options

RileyManda committed:
Removed additional print-labels

RileyManda committed:
Removed items error message when saving albums

RileyManda committed:
remobed debug messages

RileyManda committed:
Removed additional debug messages for add books

RileyManda committed:
Merge pull request #58 from RileyManda/feature/console

Feature/console

DianaBeki committed:
Fix linters

DianaBeki committed:
Fix linters

DianaBeki committed:
Fix lintewrs

DianaBeki committed:
Merge pull request #60 from RileyManda/feature/save-data-games

Save data for game class

RileyManda committed:
Added live demo to table of contents

RileyManda committed:
Added a live presentation vide link fo the project

RileyManda committed:
Updated collaborators

RileyManda committed:
Updated thank you message

RileyManda committed:
Added logo and licence to readme

RileyManda committed:
Merge pull request #61 from RileyManda/documentation

Documentation

laiifuu commented on schema.sql:
```diff
@@ -0,0 +1,69 @@
+    name VARCHAR(255) NOT NULL,
+);
+
+-- Create the Item table
+CREATE TABLE item (
+    id SERIAL PRIMARY KEY,
+    genre VARCHAR(255) NOT NULL,
+    author VARCHAR(255) NOT NULL,
+    source VARCHAR(255) NOT NULL,
+    label VARCHAR(255) NOT NULL,
+    publish_date DATE(255) NOT NULL,
+    archived BOOLEAN(255) NOT NULL,
+    can_be_archived BOOLEAN(255) NOT NULL,
+    genre_id INT REFERENCES genre(id)
+);
```
When creating SQL tables, there isn't a mechanism to portray inheritance between the tables, and so the better way to do this is to remove the item table, and add its properties to the tables of the classes that inherit from it. author, genre and label should be considered as foreign keys in the music_album, book, and game tables.
Also, can_be_archived property is not present in the class diagram as a property, so it should be removed from all tables.

laiifuu commented on schema.sql:
```diff
@@ -0,0 +1,69 @@
+
+-- create table music_album
+
+CREATE TABLE music_album (
+    id SERIAL PRIMARY KEY,
+    can_be_archived BOOLEAN(255) NOT NULL,
+    on_spotify BOOLEAN,
+    item_id INT REFERENCES item(id)
+);
```
After implementing what I mentioned in the previous comment, the music album table should look like the following: 
```
CREATE TABLE music_album (
  id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  publish_date DATE,
  on_spotify BOOLEAN,
  archived BOOLEAN,
  genre_id INT REFERENCES genre(id),
  author_id INT REFERENCES author(id),
  label_id INT REFERENCES label(id)
);
```

laiifuu commented on schema.sql:
```diff
@@ -0,0 +1,69 @@
+
+-- create table books
+
+CREATE TABLE book (
+    id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
+    publisher VARCHAR(255),
+    cover_state VARCHAR(255),
+    publish_date DATE,
+    FOREIGN KEY (id) REFERENCES item(id)
+);
```
After implementing what I mentioned in the previous comment, the books table should look like the following:
```
CREATE TABLE book (
	id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
	publisher VARCHAR(100),
	cover_state VARCHAR(100)
	publish_date  DATE
	archived Boolean
	genre_id INT REFERENCES genre(id),
  	author_id INT REFERENCES author(id),
  	label_id INT REFERENCES label(id),
);
```

laiifuu commented on schema.sql:
```diff
@@ -0,0 +1,69 @@
+
+-- create table games
+
+CREATE TABLE game (
+    id INT PRIMARY KEY,
+    last_played_at TIMESTAMP,
+    multiplayer BOOLEAN,
+    FOREIGN KEY (id) REFERENCES item(id)
+);
```
After implementing what I mentioned in the previous comment, the games table should look like the following:
```
CREATE TABLE game (
  id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY, 
  archived BOOLEAN,
  multiplayer BOOLEAN,
  publish_date DATE, 
  last_played_at DATE, 
  genre_id INT REFERENCES genre(id),
  author_id INT REFERENCES author(id),
  label_id INT REFERENCES label(id)
);
```

laiifuu commented on schema.sql:
```diff
@@ -0,0 +1,69 @@
+    id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
+    first_name VARCHAR(30) NOT NULL,
+    last_name VARCHAR(30) NOT NULL,
+    item_id INT,
+    FOREIGN KEY (item_id) REFERENCES item(id)
```
A foreign key in this case is unnecessary.

laiifuu commented on schema.sql:
```diff
@@ -0,0 +1,69 @@
+    item_id INT REFERENCES item(id)
+);
+
+ALTER TABLE item ADD CONSTRAINT unique_item UNIQUE (id);
```
Considering we're removing the Item table, this line is unnecessary.

laiifuu reviewed (changes requested):
# STATUS: CHANGES REQUIRED :recycle: 

## Hi team :wave: 

Good job so far!
There are some issues that you still need to work on, but you are almost there!

![so-close](https://user-images.githubusercontent.com/93520178/261856220-b94908ad-80bf-4dc0-b50f-e65f429c0400.gif)


## Highlights
- [x] `GitFlow` respected :heavy_check_mark: 
- [x] Green linters check :heavy_check_mark: 
- [x] All needed classes were created in separate files :heavy_check_mark: 

## Required Changes ♻️
- [ ] Good work so far on the presentation video :+1: However, when reviewing it, I noticed that not all the members of the team appeared on camera. Additionally, the video didn't include a highlight about a piece of code that each of you finds interesting or is proud of making.


- [ ] When exiting the application, and then re running it to see if the data was being stored, the data for the authors and labels was not being preserved, I'd get nothing when choosing both options to list them on screen. Kindly fix this issue.
You should also make sure that the application doesn't crash when the `json` files don't exist, I deleted the files to check for this, and it did crash: 
![image](https://github.com/RileyManda/ruby-capstone-project/assets/93520178/20b7c14d-4f16-448e-b9c3-08c80f3ef64c)



_Check the comments under the review :point_down:_

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and :computer: Happy coding :computer: 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

RileyManda committed:
Updated music_album to explicitly inherit from Item

RileyManda committed:
Update: genre class:assign the genre to the item

RileyManda committed:
Linter fix

RileyManda committed:
Remived source from item class

RileyManda committed:
Merge pull request #62 from RileyManda/project-iterations-music

Project iterations music

RileyManda committed:
Updated schema

RileyManda committed:
Updated table schema

RileyManda committed:
Removed item from scheam

RileyManda committed:
Merge pull request #63 from RileyManda/schema-iterations

Schema iterations

RileyManda committed:
Bug fix:moved load music album outside scope

RileyManda committed:
Removed redundant variable

RileyManda commented:
### Updated Iterations requested by reviewer

- updated application schema and removed item table
- updated item class and removed additional boolean check

### Bug fixes:

-  run time bug-fix when quitting application
-  Added condition to create file if not exists

Please kindly review our implementation 

Thank you 🙏🏻

RileyManda committed:
Moved games.json to root

RileyManda committed:
Added condition:Create file if not exists

RileyManda committed:
Linter error fixes

RileyManda committed:
Updated tests

RileyManda committed:
Updated tests to match changes in project

RileyManda committed:
Merge pull request #64 from RileyManda/class-iterations

Class iterations

bienvenuushindi commented on item.rb:
```diff
@@ -0,0 +1,20 @@
+class Item
```
:red_circle: [REQUIRED] As requested by the previous reviewer, please ensure that your item class incorporates the `move_to_archive()` method. For a more detailed understanding of how to implement this method, I recommend reviewing the previous reviewer's comments.

bienvenuushindi replied:
:red_circle: [REQUIRED] As mentioned by the previous reviewer, Consider explicitly writing attribute writers to ensure proper associations. Here's an example of how to do this for the genre, author, and label attributes: :arrow_down: 

```
def genre=(genre)
  @genre = genre
  genre.items.push(self) unless genre.items.include?(self)
end

def author=(author)
  @author = author
  author.items.push(self) unless author.items.include?(self)
end

def label=(label)
  @label = label
  label.items.push(self) unless label.items.include?(self)
end

```

bienvenuushindi commented on book.rb:
```diff
@@ -0,0 +1,29 @@
+    SecureRandom.uuid
+  end
+
+  def can_be_archived?
+    if @cover_state || super == 'bad'
+      false
+    else
+      true
+    end
+  end
```
:red_circle: [REQUIRED] Kindly make sure you correct this method as suggested by the previous reviewer. 

> This method should return true if parent's method returns true OR if cover_state equals to "bad", otherwise it returns false:
> 
> ```
> def can_be_archived?
>     if @cover_state == bad || super
>       true
>     else
>       false
>     end
>   end
> ```
> 
> If you want to make shorter, this will do the trick:
> 
> ```
>     def can_be_archived?
>       super || @cover_state == 'bad'
>     end
> ```

RileyManda replied:
@bienvenuushindi 
Please kindly advice on how we can submit the changes for a third review?
We have lost access to the review option on the dashboard.

bienvenuushindi replied:
@RileyManda Yesterday, I mistakenly approved the project while I was still reviewing it. I'm sorry for the inconvenience."I will re-review the project, and once I confirm that all the requested changes have been implemented, you can proceed with the submission.

RileyManda replied:
Hi @bienvenuushindi
I have updated book.rb and surrounded bad with quotes 'bad'.
Please kindly review and approve if all is in order.Because the system has already moved our team to the next project.
Thank you

bienvenuushindi commented on music_album.rb:
```diff
@@ -0,0 +1,20 @@
+
+  def can_be_archived?
+    @can_be_archived
+  end
```
:red_circle: [REQUIRED] Kindly make sure you correct this method as suggested by the previous reviewer. 

> This method should:
> 
>     * return true if parent's method returns true AND if on_spotify equals true
> 
>     * otherwise, it should return false
> 
> 
> Something like this should do the job:
> 
> ```
>     def can_be_archived?
>       super && @on_spotify == true
>     end
> ```
> 
> **Note:** The @can_be_archived property is not present in the class diagram, I suggest you remove it from all classes where it was added.

bienvenuushindi reviewed (changes requested):
## Hi team,

Good job so far!

## Highlights!
- GitFlow ✔️ 
- No linter errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive README ✔️ 

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

# Status: Required Changes ♻️⛔️
_Check the comments under the review and here 👇._


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@julie-ify` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

rabayarabu committed:
Modify book & item class

rabayarabu committed:
Modify testing files

rabayarabu committed:
Fixed linters error

rabayarabu committed:
Fixed linters error

bienvenuushindi commented on book.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def can_be_archived?
+    if @cover_state == bad || super
```
:red_circle: [Requested] In Ruby, there is a difference between bad and 'bad'. When you use bad without quotes, it is treated as a variable or identifier, whereas 'bad' with single or double quotes is a string. Keep in mind that variables and strings are distinct in Ruby, and using the appropriate one in your code is important for correct behaviour. Kindly make sure you replace bad with` 'bad'`

RileyManda replied:
Hi @bienvenuushindi 
I have updated book.rb and surrounded bad with quotes 'bad'.
Please kindly review and approve if all is in order.Because the system has already moved our team to the next project.
Thank you

RileyManda replied:
hi @bienvenuushindi 
I have update book.rb cover state to:
`if @cover_state == 'bad' || super`

bienvenuushindi reviewed (changes requested):
Hi Team , 

Good job so far!
There is just one issue that you still need to work on to go to the next project but you are almost there!
<img src="https://media0.giphy.com/media/cYaB1VycDbOyOiXcCx/giphy.gif"/>

## Highlights
May I highlight the following:

- The repo looks professional.:heavy_check_mark:
- There are no linter errors:heavy_check_mark:
- You have descriptive PR:heavy_check_mark:
- Your project looks good:heavy_check_mark:

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

RileyManda committed:
Update book.rb-changed bad to 'bad'

RileyManda committed:
Merge pull request #65 from RileyManda/book-iteration

Update book.rb-changed bad to 'bad'

RileyManda committed:
Updated book.rb if @cover_state == 'bad' || super

RileyManda committed:
Merge pull request #66 from RileyManda/update-coverstate

Updated book.rb if @cover_state == 'bad' || super

bienvenuushindi reviewed (approved):
Hi team    , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

![Charlie Brown](https://media.giphy.com/media/ZdO4NenDbsQNwUwKDB/giphy.gif)

## Highlights


- The repo looks professional.:heavy_check_mark:
- Git/GitHub flow best practices have been followed:heavy_check_mark:
- You have descriptive PR:heavy_check_mark:


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.
